tensorflow/python/grappler/auto_mixed_precision_test.py

```diff
@@ -0,0 +1,555 @@
+      y2 = _conv_pool(x2)
+      y3 = _conv_pool(x3)
+      y = array_ops.concat([y1, y2, y3], axis=3)
+      y = array_ops.identity(y)
+      optimizer = gradient_descent.GradientDescentOptimizer(learning_rate=0.01)
+      g = optimizer.compute_gradients(y, [x])
+      output = (y, g)
+
+      output_val_ref, output_val, cost_graph = self._run(output)
+      node_map = _build_node_map(cost_graph.node)
+
+      self._assert_output_fp16(node_map, 'split')
+      for suffix in [''] + ['_%i' % i for i in range(1, 6)]:
+        self._assert_output_fp16(node_map, 'Conv2D' + suffix)
+        self._assert_output_fp16(node_map, 'Relu' + suffix)
+        self._assert_output_fp16(node_map, 'MaxPool' + suffix)
+      self._assert_output_fp16(node_map, 'concat')
+      self.assertAllClose(output_val_ref, output_val, atol=1e-3, rtol=1e-3)
+
+  def test_multi_paths_2(self):
+    """ Test graph with multiple paths """
+    if test.is_gpu_available(cuda_only=True,
+                             min_cuda_compute_capability=self.MIN_GPU_ARCH):
+      random_seed.set_random_seed(0)
+      x = _input([8, 8])
+      y1 = _matmul_act(x)
+      y2 = _matmul_act(x)
+      y = y1 + y2 + x
+      optimizer = gradient_descent.GradientDescentOptimizer(learning_rate=0.01)
+      g = optimizer.compute_gradients(y, [x])
+      output = (g, y)
+
+      output_val_ref, output_val, cost_graph = self._run(output)
+      node_map = _build_node_map(cost_graph.node)
+
+      self._assert_output_fp16(node_map, 'MatMul')
+      self._assert_output_fp16(node_map, 'Relu')
+      self._assert_output_fp16(node_map, 'MatMul_1')
+      self._assert_output_fp16(node_map, 'Relu_1')
+      self.assertAllClose(output_val_ref, output_val, atol=1e-3, rtol=1e-3)
+
+  def test_recurrent_lstm(self):
+    """ Test graph with recurrent lstm """
+    if test.is_gpu_available(cuda_only=True,
+                             min_cuda_compute_capability=self.MIN_GPU_ARCH):
+      random_seed.set_random_seed(0)
+      init_c = _input([8, 4])
+      init_h = _input([8, 4])
+      _, _, h, _ = _recurrent_lstm(init_c, init_h)
+      optimizer = gradient_descent.GradientDescentOptimizer(learning_rate=0.01)
+      g = optimizer.compute_gradients(h, [init_c, init_h])
+      output = (h, g)
+
+      output_val_ref, output_val, cost_graph = self._run(output)
+      node_map = _build_node_map(cost_graph.node)
+
+      self._assert_output_fp16(node_map, 'while/concat')
+      self._assert_output_fp16(node_map, 'while/MatMul')
+      self._assert_output_fp16(node_map, 'while/split')
+      self._assert_output_fp16(node_map, 'while/Sigmoid')
+      self._assert_output_fp16(node_map, 'while/Sigmoid_1')
+      self._assert_output_fp16(node_map, 'while/Sigmoid_2')
+      self._assert_output_fp16(node_map, 'while/Tanh')
+      self._assert_output_fp16(node_map, 'while/Tanh_1')
+      self.assertAllClose(output_val_ref, output_val, atol=1e-3, rtol=1e-3)
+
+  def test_propagation_through_intertwined_loop_1(self):
```

Comment:
I think having a few loop tests that are intertwined would be nice. We have one but it is very simple. I am worried someone will break these tests, but not any others, and be confused about why the expected dtypes are these values. Adding a TODO is fine.

Reply:
Given your comment about intertwined loops and the fact that the previous tests were mostly producing NaNs, I've replaced them with a set of simple loop tests.